src/bootloaders/shim-systemd.c

```diff
@@ -214,57 +223,46 @@ static bool shim_systemd_needs_update(__cbm_unused__ const BootManager *manager)
 static bool make_layout(const BootManager *manager)
 {
         char *boot_root = boot_manager_get_boot_dir((BootManager *)manager);
-        char *path = NULL;
-        path = string_printf("%s%s", boot_root, DST_DIR);
-        if (!nc_mkdir_p(path, 00755)) {
-                goto fail;
-        }
-        free(path);
-        path = string_printf("%s%s", boot_root, KERNEL_DST_DIR);
-        if (!nc_mkdir_p(path, 00755)) {
+        char *systemd_config_entries = NULL;
+
+        if (!nc_mkdir_p(kernel_dst_host, 00755)) {
                 goto fail;
         }
-        free(path);
-        path = string_printf("%s%s", boot_root, SYSTEMD_ENTRIES);
-        if (!nc_mkdir_p(path, 00755)) {
+
+        systemd_config_entries = string_printf("%s/" SYSTEMD_CONFIG_DIR "/" SYSTEMD_ENTRIES_DIR, boot_root);
+        if (!nc_mkdir_p(systemd_config_entries, 00755)) {
                 goto fail;
         }
-        free(path);
         /* in case of image creation, override the fallback bootloader, so the
          * media will be bootable. */
         if (is_image_mode) {
-                path = string_printf("%s%s", boot_root, EFI_FALLBACK_DIR);
-                LOG_INFO("Image mode: creating dir for fallback: %s", path);
-                if (!nc_mkdir_p(path, 00755)) {
+                char *efi_fallback_dir = dirname(strdup(efi_fallback_dst_host));
```

Comment:
strdup can fail, so can dirname.

Reply:
Yes, I'll make `get_kernel_destination()` return full path and get rid of all this ugliness. Sorry.